### ISSUE TITLE
Artifacts in bicubic interpolation

### ISSUE THREAD

**Issue by mattwigway** [conveyal/gridualizer#21]
We're seeing some artifacts in bicubic interpolation:

![image](https://cloud.githubusercontent.com/assets/566958/24105401/2008e938-0d5b-11e7-9412-a661a3d4deaf.png)

**Comment by mattwigway** [comment]
Per conversation on Slack, we believe this is due to overshoots from the bicubic interpolator, in situations like these:

![image](https://cloud.githubusercontent.com/assets/566958/24105666/fae163a0-0d5b-11e7-8908-1960e9e43188.png)

@abyrd points out that this not necessarily wrong (it's possible that the local minimum occurs between samples), but is clearly incorrect when the interpolator is jumping down very far in order to create the steep slope needed to reach hard edges. @mattwigway is additionally concerned about the optics of having floating local minima because there could be a transit stop in the middle of them.

**Comment by mattwigway** [comment]
There is https://en.wikipedia.org/wiki/Monotone_cubic_interpolation but our values are actually no monotonic, we just want to constrain them somewhat. Equation 4.4 of [Dougherty, Randall L, Alan S Edelman, and James M Hyman. “Nonnegativity-, Monotonicity-, or Convexity-Preserving Cubic and Quintic Hermite Interpolation.” Mathematics of Computation 52, no. 186 (1989): 471–94. doi:S0025-5718-1989-0962209-1.](http://www.ams.org/mcom/1989-52-186/S0025-5718-1989-0962209-1/S0025-5718-1989-0962209-1.pdf) seems promising.

**Comment by mattwigway** [comment]
Specifically, strict local monotonicity can create situations such as these (image from above paper): 
![image](https://cloud.githubusercontent.com/assets/566958/24107757/00edc602-0d62-11e7-9889-367a2be1195d.png)

**Comment by abyrd** [comment]
http://www.korf.co.uk/spline.pdf

**Comment by mattwigway** [comment]
I would be in favor of the spline method. It seems to be relatively simple, and I think that while there is some theoretical justification for the overshoots, for the isochrones anyhow we're going to be better off with something that isn't too clever. The isochrones are already an oversimplification (since they show an exact boundary, when we have in fact computed a field), so I think the guiding principle here should be that of least astonishment; the isochrones should follow the values we computed, and beyond that an interpolation method should be chosen based on what "looks right."

**Comment by abyrd** [comment]
Agreed, though the overshoots are technically justified, they are clearly undesirable and are the result of insufficient constraints on the interpolation process. I do want to first try clamping the input values and/or giving special treatment to 255 (unreachable) values, just as an experiment to see if that's what triggers most of the overshoot artifacts. But if any remain at all, there's no reason to keep using the bicubic interpolator. There are clearly other methods specifically designed to avoid overshoot, like the above-linked Constrained Cubic Spline Interpolation for Chemical Engineering Applications by CJC Kruger.

**Comment by abyrd** [comment]
This should be solved by PR #24
